### PR TITLE
Nest 'introduce variable' code actions to help clean up the lightbulb.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Makes things now look like this:

image

Note: we will inline these guys if the lightbulb is short.  For example:

image

This fits the IDE style of trying to balance between providing fast access to helpful stuff, while also not making the code action list hugely overwhelming.  

This came up as part of the discussion during: https://github.com/dotnet/roslyn/issues/30464

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @jinujoseph @sharwell @Neme12 @jcouv

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide Very simple PR if someone wants to take a look.

**Comment by CyrusNajmabadi** [comment]
> I'm not seeing any tests for the new behavior. 

Sure.  I'll make it more explicit :)

**Comment by CyrusNajmabadi** [comment]
Ok.  I've updated the testing harness.  It now validates that the fixes it tests *must* have been nested.  So, if that changes, the tests will blow up.

**Comment by CyrusNajmabadi** [comment]
@sharwell For re-review.

**Comment by CyrusNajmabadi** [comment]
Note: i don't want to make this change now, but i feel that we should probably make "extract method" have higher priority than "introduce variable".  There are two reasons i want this:

1. i actually think EM is used more than IV.
2. EM only has one action.  So if it's not the right one, you only need to skip over one thing to get to the IV actions.  If you don't want the IV action, you may need to skip over up to actions to get the EM action.

It would be great if we looked at the numbers for each and prioritized accordingly.  

@sharwell Is this something you'd be interested in looking at?  Previously it was Kasey.  But i don't know who has taken over that role now.

**Comment by sharwell** [comment]
> It now validates that the fixes it tests must have been nested

💯

**Comment by CyrusNajmabadi** [comment]
@jcouv @dpoeschl can you look at this fairly simple PR?

**Comment by jcouv** [comment]
Looking

File: src/Features/Core/Portable/FeaturesResources.resx [hunk 1]
```diff
@@ -1451,4 +1451,16 @@ This version used in: {2}</value>
...
+  <data name="Introduce_constant" xml:space="preserve">
+    <value>Introduce constant</value>
+  </data>
+  <data name="Introduce_field" xml:space="preserve">
+    <value>Introduce field</value>
+  </data>
+  <data name="Introduce_local" xml:space="preserve">
+    <value>Introduce local</value>
+  </data>
+  <data name="Introduce_query_variable" xml:space="preserve">
```
Comment (jcouv) [inline src/Features/Core/Portable/FeaturesResources.resx]: >Introduce_query_variable [](start = 14, length = 24)

Would you mind sharing a screenshot of this scenario? I don't think I've encountered it before.

File: src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs [hunk 2]
```diff
@@ -61,50 +63,74 @@ protected virtual bool BlockOverlapsHiddenPosition(SyntaxNode block, Cancellatio
...
+                    if (actions.Length > 0)
                     {
-                        return actions.AsImmutableOrNull();
+                        // We may end up creating a lot of viable code actions for the selected
+                        // piece of code.  Create a top level code action so that we don't overwhelm
+                        // the light bulb if there are a lot of other options in the list.  Set 
+                        // the code action as 'inlinable' so that if the lightbulb is not cluttered
+                        // then the nested items can just be lifted into it, giving the user fast
+                        // access to them.
+                        return new CodeActionWithNestedActions(title, actions, isInlinable: true);
```
Comment (jcouv) [inline src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs]: >title [](start = 63, length = 5)

Just to confirm: this title string which is passed around is subject to localization, right?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs]: Correct.

File: src/Features/Core/Portable/FeaturesResources.resx (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/FeaturesResources.resx]: Sure!  Looks like:

image

**Comment by CyrusNajmabadi** [comment]
@jcouv feel free to merge in.

**Comment by Neme12** [comment]
There are failing tests.

**Comment by CyrusNajmabadi** [comment]
Oh interesting.  THe testing system has changed.  I didn't realize that :)

**Comment by CyrusNajmabadi** [comment]
@jcouv have addressed tests.  sorry about that :)

**Comment by jcouv** [comment]
Merged (I assume you still don't like squashing). Thanks!

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**